### ISSUE TITLE
Vendoring lalrpop-* causes generated files to be stored as part of the mirror

### ISSUE THREAD

**Issue by Yoric** [lalrpop/lalrpop#333]
I'm working on a change to the Firefox build system that indirectly depends on crates lalrpop and lalrpop-snap. Unfortunately, when calling `cargo vendor`, this causes the following files to be stored:

lalrpop-snap/src/newparser.rs
lalrpop-snap/src/parser/lrgrammar.rs
lalrpop/src/newparser.rs
lalrpop/src/parser/lrgrammar.rs

As far as I can tell, these files are generated by build.rs, so should not be stored.

This is a blocker for [Firefox bug 1437004](https://bugzilla.mozilla.org/show_bug.cgi?id=1437004.)

**Comment by mbrubeck** [comment]
It looks like this was fixed by #295 which has not yet been published to crates.io.

**Comment by Yoric** [comment]
Would it be possible to backport that patch and/or publish a new version on crates.io?

**Comment by Yoric** [comment]
This is a blocker for https://bugzilla.mozilla.org/show_bug.cgi?id=1437004. @nikomatsakis, what's the best way to get #295 released quickly?

**Comment by Marwes** [comment]
> lalrpop-snap/src/parser/lrgrammar.rs (1.9Mb)

This file won't be possible to remove since it must be kept as-is in `lalrpop-snap` (there is no way to generate it since `lalrpop-snap` must be a complete working copy of lalrpop).

> lalrpop/src/parser/lrgrammar.rs (3.3Mb)

I think this one could be removed with some hacks. Essentially we would generated it into `OUT_DIR` for the `lalrpop` crate and use it as explained in #280 . Then in snap we could with a feature flag change it to look for the file at `parser/src/lrgrammar.rs` instead of looking in `OUT_DIR` (so that snap is still self contained). 

The only problem is that I don't have a good idea on how to make https://github.com/lalrpop/lalrpop/blob/master/snap.sh find `OUT_DIR` and copy the parser from there.

@nikomatsakis I can draft up a release (like https://github.com/lalrpop/lalrpop/pull/288) tomorrow if you want. I can't publish however.

**Comment by Marwes** [comment]
I should also mention that one idea I had for reducing binary size in #304 would also reduce the size of the generated file quite a bit as well

> Another improvement which can help in some cases (if it is possible). Is to avoid generating unique tables for each pub item. Gluon exports two entrypoints and these each get their own unique tables despite being almost the same. I think it is possible to generate a single parser for both of these and just generate a small initializer for each which starts the parser of in the correct state after which everything should work as normal (please correct me if I am wrong about this @nikomatsakis ).

I have started to look into that a bit and it seems plausible though not without a bunch of refactoring. Since lalrpop has 5 public parsers there is quite a lot of potential for reduction through that.

**Comment by Yoric** [comment]
Well, any improvement here would be helpful. Do you have an idea of when all of this could land?

**Comment by Marwes** [comment]
> Well, any improvement here would be helpful. Do you have an idea of when all of this could land?

Afraid not, https://github.com/lalrpop/lalrpop/issues/333#issuecomment-373947237 is easy enough to do (though it is a bit of a hack :/). But as long as `snap.sh` can't copy the generated parser there is no way to implement it.

https://github.com/lalrpop/lalrpop/issues/333#issuecomment-373947729 I can't say either. I'd like to get the binary size decrease from it myself but it is more of a nice to have than a necessity so it is currently of medium/low priority for me.

**Comment by Yoric** [comment]
I'm trying to understand why vendoring a crate that depends from lalrpop causes lalrpop-snap to be also vendored in.

**Comment by Marwes** [comment]
`lalrpop-snap` is needed to generate the parser used in the `lalrpop` crate. In theory we could generate this parser into the `lalrpop` crate before publishing it which would make it possible to drop the `lalrpop-snap` dependency on the build *published* to crates.io. However this would mean doing some rather hacky and brittle things to the `Cargo.toml` file just before publishing and then reverting afterwards (among other things).

**Comment by Yoric** [comment]
Ok, I just encountered this problem again when updating dependencies for Firefox.

Blocked by lrgrammar.rs (2.9Mb).

**Comment by Marwes** [comment]
Got some PRs up for review now which works on reducing the size of lalrpop (in multiple respects) the final tally is a 1Mb .rs file for the grammar https://github.com/lalrpop/lalrpop/pull/414 . Reducing it further is possible but it is certainly getting more difficult and it is unclear whether the reductions will translate into binary size savings (which is what I mostly care about for these changes).

**Comment by Marwes** [comment]
0.16.1 is now released which is includes the reductions to lrgrammar.rs size.

**Comment by nox** [comment]
Couldn't we generate a file that contains the lalrpop grammar as its AST, rather than the result of the compilation of that AST to Rust code?

**Comment by Marwes** [comment]
That isn't really possible as any change in the type definitions used in the generated AST could break compilation.

**Comment by Marwes** [comment]
If the current size (1Mb) is still an issue, there are still ways to get several 5-10% reductions fairly easily but we are at the point of diminishing returns now.

To put things in perspective, the rest of the `lalrpop` crate is about 1.2Mb in size (so 2.2Mb in total) so reducing the size of `lrgrammar.rs` could at best cut the size cost (in git) of including `lalrpop` in half.
